src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -50,153 +51,137 @@ case class InferReadWriteAnnotation(t: String, tID: TransID)
-              WRef("clk", ClockType, NodeKind, MALE))
-            stmts += Connect(NoInfo, WSubField(rw_exp, "en", bt, FEMALE),
-              DoPrim(Or, List(connects(s"${mem.name}.$r.en"), connects(s"${mem.name}.$w.en")), Nil, bt))
-            stmts += Connect(NoInfo, WSubField(rw_exp, "addr", ut, FEMALE),
-              Mux(connects(s"${mem.name}.$w.en"), connects(s"${mem.name}.$w.addr"), 
-                  connects(s"${mem.name}.$r.addr"), ut))
-          }
-        }
-        if (readwriters.isEmpty) mem else DefMemory(mem.info,
-          mem.name, mem.dataType, mem.depth, mem.writeLatency, mem.readLatency,
-          mem.readers filterNot readers, mem.writers filterNot writers,
-          mem.readwriters ++ readwriters)
+  def replaceStmt(repl: Netlist)(s: Statement): Statement =
+    s map replaceStmt(repl) map replaceExp(repl) match {
+      case Connect(_, EmptyExpression, _) => EmptyStmt 
       case s => s
     }
-
-    def replaceExp(e: Expression): Expression =
-      e map replaceExp match {
-        case e: WSubField => repl getOrElse (e.serialize, e)
-        case e => e
+    
+  def inferReadWriteStmt(connects: Connects,
+                         repl: Netlist,
+                         stmts: Statements)
+                         (s: Statement): Statement = s match {
+    // infer readwrite ports only for non combinational memories
+    case mem: DefMemory if mem.readLatency > 0 =>
+      val ut = UnknownType
+      val ug = UNKNOWNGENDER
+      val readers = new PortSet
+      val writers = new PortSet
+      val readwriters = collection.mutable.ArrayBuffer[String]()
+      val namespace = Namespace(mem.readers ++ mem.writers ++ mem.readwriters)
+      for (w <- mem.writers ; r <- mem.readers) {
+        val wp = getProductTerms(connects)(memPortField(mem, w, "en"))
+        val rp = getProductTerms(connects)(memPortField(mem, r, "en"))
+        if (wp exists (a => rp exists (b => checkComplement(a, b)))) {
+          val rw = namespace newName "rw"
+          val rwExp = WSubField(WRef(mem.name, ut, MemKind, ug), rw, ut, ug)
+          readwriters += rw
+          readers += r
+          writers += w
+          repl(memPortField(mem, r, "en"))   = EmptyExpression
+          repl(memPortField(mem, r, "clk"))  = EmptyExpression
+          repl(memPortField(mem, r, "addr")) = EmptyExpression
+          repl(memPortField(mem, r, "data")) = WSubField(rwExp, "rdata", mem.dataType, MALE)
+          repl(memPortField(mem, w, "en"))   = WSubField(rwExp, "wmode", BoolType, FEMALE)
+          repl(memPortField(mem, w, "clk"))  = EmptyExpression
+          repl(memPortField(mem, w, "addr")) = EmptyExpression
+          repl(memPortField(mem, w, "data")) = WSubField(rwExp, "wdata", mem.dataType, FEMALE)
+          repl(memPortField(mem, w, "mask")) = WSubField(rwExp, "wmask", ut, FEMALE)
+          stmts += Connect(NoInfo, WSubField(rwExp, "clk", ClockType, FEMALE),
+            WRef("clk", ClockType, NodeKind, MALE))
```

Comment:
Does this work when the memory clock isn't "clk"? i.e. the clk going into the top level module is named something different....